File: javascript/packages/highlighter/test/html-diagnostic.test.ts
import { describe, test, expect, beforeAll, beforeEach } from "vitest"

import { Herb } from "@herb-tools/node-wasm"
import { Highlighter } from "../src/highlighter.js"
import { renderDocumentHTML, wrapDocumentHTML } from "../src/html-sink.js"
import { generateStylesheet } from "../src/stylesheet.js"
import { themes } from "../src/themes.js"

import type { Diagnostic } from "@herb-tools/core"

const PATH = "app/views/users/show.html.erb"
const CONTENT = "<div>\n  <%= user.name %>\n  <p>hello</p>\n  <span>world</span>\n</div>"

const DIAG_MULTI: Diagnostic = {
  severity: "error",
  location: {
    start: { line: 2, column: 2 },
    end: { line: 3, column: 6 },
  },
  message: "unclosed `<%=` tag",
  code: "parser-error",
}

const DIAG_WARN: Diagnostic = {
  severity: "warning",
  location: {
    start: { line: 4, column: 2 },
    end: { line: 4, column: 8 },
  },
  message: "avoid inline spans",
}

const OVERLAP_CONTENT = "<div id=\"x\">text</div>"

const DIAG_OVERLAP_ERROR: Diagnostic = {
  severity: "error",
  location: {
    start: { line: 1, column: 5 },
    end: { line: 1, column: 12 },
  },
  message: "first",
}

const DIAG_OVERLAP_WARNING: Diagnostic = {
  severity: "warning",
  location: {
    start: { line: 1, column: 8 },
    end: { line: 1, column: 16 },
  },
  message: "second",
}

const DIAG_UNSAFE: Diagnostic = {
  severity: "error",
  location: {
    start: { line: 2, column: 2 },
    end: { line: 2, column: 6 },
  },
  message: "bad <img> in `attr`",
  code: "unsafe-rule",
}

describe("HTMLDiagnostic", () => {
  let highlighter: Highlighter

  beforeAll(async () => {
    await Herb.load()
  })

  beforeEach(async () => {
    highlighter = new Highlighter("onedark")
    await highlighter.initialize()
  })

  test("renders an inline diagnostics document with span markers", () => {
    const document = highlighter.buildDocument(PATH, CONTENT, { diagnostics: [DIAG_MULTI, DIAG_WARN] })
    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "spans" })

    expect(result).toMatchSnapshot()
  })

  test("renders an inline diagnostics document with highlight api markers", () => {
    const document = highlighter.buildDocument(PATH, CONTENT, { diagnostics: [DIAG_MULTI, DIAG_WARN] })
    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "highlight-api" })

    expect(result).toMatchSnapshot()
  })

  test("renders a card document", () => {
    const document = highlighter.buildCard(PATH, DIAG_MULTI, CONTENT, {
      contextLines: 1,
      optimizeHighlighting: true,
      codeUrl: "https://herb.tools/rules/parser-error",
      fileUrl: "file:///app/views/show.html.erb",
      suffix: "(fixable)",
    })

    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "spans" })

    expect(result).toMatchSnapshot()
  })

  test("renders a split document with progress rules", () => {
    const document = highlighter.buildDocument(PATH, CONTENT, {
      diagnostics: [DIAG_MULTI, DIAG_WARN],
      splitDiagnostics: true,
      contextLines: 1,
    })

    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "spans" })

    expect(result).toMatchSnapshot()
  })

  test("escapes markup in messages and rejects unsafe urls", () => {
    const document = highlighter.buildDocument(PATH, CONTENT, {
      diagnostics: [DIAG_UNSAFE],
      codeUrlBuilder: () => "javascript:alert(1)",
    })

    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "spans" })

    expect(result).toMatchSnapshot()
  })

  test("overlapping markers use the highest severity", () => {
    const document = highlighter.buildDocument(PATH, OVERLAP_CONTENT, {
      diagnostics: [DIAG_OVERLAP_ERROR, DIAG_OVERLAP_WARNING],
    })

    const result = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "spans" })

    expect(result).toMatchSnapshot()
  })

  test("wraps a document with chrome and hydration", () => {
    const document = highlighter.buildDocument(PATH, CONTENT, { diagnostics: [DIAG_MULTI, DIAG_WARN] })
    const fragment = renderDocumentHTML(document, { themeLabel: "onedark", showLineNumbers: true, markers: "highlight-api" })
    const result = wrapDocumentHTML(fragment, generateStylesheet(themes.onedark, "onedark"), true)

    expect(result).toMatchSnapshot()
  })
})

File: rust/herb-highlighter/tests/html_diagnostic_test.rs
use herb_highlighter::diagnostic::{Diagnostic, DiagnosticSeverity, SerializedLocation};
use herb_highlighter::document_builder::{CardOptions, DocumentBuilder};
use herb_highlighter::highlighter::{HighlightOptions, Highlighter};
use herb_highlighter::html_sink::{render_document_html, wrap_document_html, HTMLSinkOptions, MarkerMode};
use herb_highlighter::stylesheet::generate_stylesheet;
use herb_highlighter::syntax_renderer::SyntaxRenderer;
use herb_highlighter::themes::{get_theme, Theme};

const PATH: &str = "app/views/users/show.html.erb";
const CONTENT: &str = "<div>\n  <%= user.name %>\n  <p>hello</p>\n  <span>world</span>\n</div>";
const OVERLAP_CONTENT: &str = "<div id=\"x\">text</div>";

fn multi_line_diagnostic() -> Diagnostic {
  Diagnostic::new("unclosed `<%=` tag", SerializedLocation::from(2, 2, 3, 6), DiagnosticSeverity::Error).with_code("parser-error")
}

fn warning_diagnostic() -> Diagnostic {
  Diagnostic::new("avoid inline spans", SerializedLocation::from(4, 2, 4, 8), DiagnosticSeverity::Warning)
}

fn overlap_error_diagnostic() -> Diagnostic {
  Diagnostic::new("first", SerializedLocation::from(1, 5, 1, 12), DiagnosticSeverity::Error)
}

fn overlap_warning_diagnostic() -> Diagnostic {
  Diagnostic::new("second", SerializedLocation::from(1, 8, 1, 16), DiagnosticSeverity::Warning)
}

fn unsafe_diagnostic() -> Diagnostic {
  Diagnostic::new("bad <img> in `attr`", SerializedLocation::from(2, 2, 2, 6), DiagnosticSeverity::Error).with_code("unsafe-rule")
}

fn highlighter() -> Highlighter {
  Highlighter::new("onedark").expect("the theme resolves")
}

fn sink_options(markers: MarkerMode) -> HTMLSinkOptions {
  HTMLSinkOptions {
    theme_label: "onedark".to_string(),
    show_line_numbers: true,
    markers,
  }
}

#[test]
fn renders_an_inline_diagnostics_document_with_span_markers() {
  let diagnostics = [multi_line_diagnostic(), warning_diagnostic()];

  let document = highlighter().build_document(
    PATH,
    CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::Spans)));
}

#[test]
fn renders_an_inline_diagnostics_document_with_highlight_api_markers() {
  let diagnostics = [multi_line_diagnostic(), warning_diagnostic()];

  let document = highlighter().build_document(
    PATH,
    CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::HighlightApi)));
}

#[test]
fn renders_a_card_document() {
  let syntax_renderer = SyntaxRenderer::new(get_theme(Theme::OneDark).clone());
  let builder = DocumentBuilder::new(&syntax_renderer);

  let document = builder.build_card(
    PATH,
    &multi_line_diagnostic(),
    CONTENT,
    &CardOptions {
      context_lines: 1,
      optimize_highlighting: true,
      code_url: Some("https://herb.tools/rules/parser-error".to_string()),
      file_url: Some("file:///app/views/show.html.erb".to_string()),
      suffix: Some("(fixable)".to_string()),
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::Spans)));
}

#[test]
fn renders_a_split_document_with_progress_rules() {
  let diagnostics = [multi_line_diagnostic(), warning_diagnostic()];

  let document = highlighter().build_document(
    PATH,
    CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      split_diagnostics: true,
      context_lines: 1,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::Spans)));
}

#[test]
fn escapes_markup_in_messages_and_rejects_unsafe_urls() {
  let diagnostics = [unsafe_diagnostic()];
  let code_url_builder = |_code: &str| "javascript:alert(1)".to_string();

  let document = highlighter().build_document(
    PATH,
    CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      code_url_builder: Some(&code_url_builder),
      ..Default::default()
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::Spans)));
}

#[test]
fn overlapping_markers_use_the_highest_severity() {
  let diagnostics = [overlap_error_diagnostic(), overlap_warning_diagnostic()];

  let document = highlighter().build_document(
    PATH,
    OVERLAP_CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      ..Default::default()
    },
  );

  insta::assert_snapshot!(render_document_html(&document, &sink_options(MarkerMode::Spans)));
}

#[test]
fn wraps_a_document_with_chrome_and_hydration() {
  let diagnostics = [multi_line_diagnostic(), warning_diagnostic()];

  let document = highlighter().build_document(
    PATH,
    CONTENT,
    &HighlightOptions {
      diagnostics: &diagnostics,
      ..Default::default()
    },
  );

  let fragment = render_document_html(&document, &sink_options(MarkerMode::HighlightApi));
  let stylesheet = generate_stylesheet(get_theme(Theme::OneDark), "onedark", None);

  insta::assert_snapshot!(wrap_document_html(&fragment, &stylesheet, true));
}
